cou/apps/base.py

```diff
@@ -794,13 +794,28 @@ def _check_application_target(self, target: OpenStackRelease) -> None:
         if (
             self.current_os_release >= target
             and not self.can_upgrade_to
-            and self.apt_source_codename >= target
+            and not self._need_origin_setting_update(target)
```

Comment:
while this should work, I think a better solution would be altering how `apt_source_codename` behaves. That property is effectively answering the question: "if I were to run `apt upgrade` on this unit, what OpenStack release would my (potentially updated) software be compatible with?"; for units with an origin config it's either parsing the cloudarchive string or resolving series → OpenStack release because the cloudarchive string is converted into a new source entry (and "distro" is not), but why not seeing units without such setting in a similar way? If there's no origin config it simply means that it's not possible to override the unit sources (i.e. the "distro" condition), so `apt_source_codename` for vault or units with an empty origin could be whatever OpenStack release the unit series is compatible with.

[sidenote: if we were to go ahead with this PR as is, line 791 should be updated as well (since `apt_source_codename` can be `None`)]

Reply:
I think I see your point and I like the idea. However this is a configuration on the application level and not into the unit. I think a safe path here would be considering the min version on the units as the `apt_source_codename`

Reply:
ftr I was using "unit" as a substitute for "machine, VM, or LXD", meaning simply "an Ubuntu installation"